### PR TITLE
Support MCT transfers

### PR DESCRIPTION
**Pull request by jameswhollister**
## Ticket
* 44372bda5b & 2bb9a38eb5

## Changes
* Support MCT transfers
* Improve validation for transfers of tokens with different decimals

### CODE REVIEW

File: src/renderer/services/dex.js [hunk 1]
```diff
@@ -619,10 +624,26 @@ export default {
           depositPromises.push(this.depositAsset(d.assetId, d.quantityToDeposit.toNumber()));
         });
 
+        const watchInterval = setInterval(() => {
+          let waiting = false;
+          order.deposits.forEach((d) => {
```
Comment (crishellco) [inline src/renderer/services/dex.js]: Those darned single letter variables...

File: src/renderer/services/dex.js [hunk 2]
```diff
@@ -788,59 +809,84 @@ export default {
...
-                .catch((e) => {
-                  reject(e);
-                });
-            } else {
-              reject('Transaction rejected');
-            }
+                    resolve(res.tx);
+                  })
+                  .catch((e) => {
+                    reject(e);
```
Comment (crishellco) [inline src/renderer/services/dex.js]: Explicit message, not exception object.

File: src/renderer/services/dex.js [hunk 3]
```diff
@@ -619,10 +624,26 @@ export default {
...
+            clearInterval(watchInterval);
+            resolve(order);
+          }
+        }, intervals.BLOCK);
+
         Promise.all(depositPromises)
           .then(() => {
             store.dispatch('fetchHoldings', {
               done: () => {
+                clearInterval(watchInterval);
```
Comment (jsolman) [inline src/renderer/services/dex.js]: Is this redundant with the clearInterval called when on line 637?

File: src/renderer/services/dex.js [hunk 4]
```diff
@@ -788,59 +809,84 @@ export default {
...
-            if (res.success) {
-              alerts.success('Deposit relayed, waiting for confirmation...');
-              neo.monitorTransactionConfirmation(res.tx, true)
-                .then(() => {
-                  const inMemoryHolding = _.get(store.state.nep5Balances, assetId);
-                  if (inMemoryHolding) {
-                    inMemoryHolding.balance = null;
-                  }
+        if (holding.canPull !== false) {
+          this.executeContractTransaction('deposit',
```
Comment (jsolman) [inline src/renderer/services/dex.js]: We should break the code inside the two branches of this if into methods.

File: src/renderer/components/dashboard/Send.vue [hunk 5]
```diff
@@ -148,17 +149,27 @@ export default {
...
+        if (this.currency) {
+          cleanAmount = new BigNumber(cleanAmount).toFixed(this.currency.decimals != null ? this.currency.decimals : 8);
         } else if (cleanAmount[cleanAmount.length - 1] !== '.'
           && cleanAmount[cleanAmount.length - 1] !== '0') {
           const n = new BigNumber(cleanAmount);
           cleanAmount = this.$formatNumber(n, this.$constants.formats.WHOLE_NUMBER_NO_COMMAS);
         }
       }
 
+      if (cleanAmount.indexOf('.') > -1) {
```
[the remark this answers is not in the document]
Comment (chovy) [reply src/renderer/components/dashboard/Send.vue]: +1

File: src/renderer/components/modals/OrderConfirmationModal.vue [hunk 6]
```diff
@@ -52,16 +52,16 @@
...
+        </div>
+        <div v-if="$store.state.orderToConfirm.deposits.length > 0">
+          <div class="deposit" v-for="(deposit, index) in $store.state.orderToConfirm.deposits" :key="index">
+            {{$t('thisOrderRequires', {
+            quantity: $formatNumber(deposit.quantityRequired),
+            symbol: deposit.symbol,
+            balance: $formatNumber(deposit.currentQuantity),
+            deposit: $formatNumber(deposit.quantityToDeposit),
+            })
+            }}
```
Comment (chovy) [inline src/renderer/components/modals/OrderConfirmationModal.vue]: should indent here so you have:

```
    })
})
```

File: src/renderer/services/dex.js [hunk 7]
```diff
@@ -788,59 +809,84 @@ export default {
...
+            [
+              u.reverseHex(assetId),
+              u.num2fixed8(quantity),
+            ], neoToSend, gasToSend)
+            .then((res) => {
+              if (res.success) {
+                alerts.success('Deposit relayed, waiting for confirmation...');
+                neo.monitorTransactionConfirmation(res.tx, true)
```
Comment (chovy) [inline src/renderer/services/dex.js]: This should use l10n/i18n: `this.$t('depositRelayedWaiting')` and add the key to en.js

File: src/renderer/services/dex.js [hunk 8]
```diff
@@ -788,59 +809,84 @@ export default {
...
-                .catch((e) => {
-                  reject(e);
-                });
-            } else {
-              reject('Transaction rejected');
-            }
+                    resolve(res.tx);
+                  })
+                  .catch((e) => {
+                    reject(`Deposit Failed. ${e.message}`);
```
Comment (chovy) [inline src/renderer/services/dex.js]: Same here with i18n; there are a few other spots.

File: src/renderer/services/dex.js [hunk 9]
```diff
@@ -788,59 +809,84 @@ export default {
...
+                  });
+              } else {
+                reject('Transaction rejected');
```
Comment (chovy) [inline src/renderer/services/dex.js]: `this.$t(...)`

File: src/renderer/services/dex.js [hunk 10]
```diff
@@ -788,59 +809,84 @@ export default {
...
+                inMemoryHolding.balance = null;
+              }
+            })
+            .catch((e) => {
+              reject(`Deposit Failed. ${e.message}`);
+            });
+        } else {
+          const dexAddress = wallet.getAddressFromScriptHash(assets.DEX_SCRIPT_HASH);
+          neo.sendFunds(dexAddress, assetId, quantity, true, () => {
+            alerts.success('Deposit relayed, waiting for confirmation...');
```
Comment (chovy) [inline src/renderer/services/dex.js]: this.$t

File: src/renderer/services/dex.js [hunk 11]
```diff
@@ -788,59 +809,84 @@ export default {
...
+
+              // set in memory holding balance to null so it will pick up the new balance
+              // if it was skipping it before because we didn't hold any
+              const inMemoryHolding = _.get(store.state.nep5Balances, assetId);
+              if (inMemoryHolding) {
+                inMemoryHolding.balance = null;
+              }
+            })
+            .catch((e) => {
+              reject(`Deposit Failed. ${e.message}`);
```
Comment (chovy) [inline src/renderer/services/dex.js]: this.$t

File: src/renderer/services/dex.js [hunk 12]
```diff
@@ -788,59 +809,84 @@ export default {
...
           gasToSend = quantity;
-          if (gasToSend > gasHolding.balance) {
+          if (gasToSend > holding.balance) {
             reject('Insufficient GAS.');
             return;
           }
+        } else {
+          holding = neo.getHolding(assetId);
+          if (holding.balance.isLessThan(quantity)) {
+            reject(`Insufficient ${holding != null ? holding.symbol : assetId}.`);
```
Comment (chovy) [inline src/renderer/services/dex.js]: use `this.$t`

File: src/renderer/services/dex.js (hunk 3 above)
Comment (jameswhollister) [reply src/renderer/services/dex.js]: No they're both necessary, the one lower may find the deposit succeeded first. The top is more of a failsafe

File: src/renderer/services/dex.js (hunk 8 above)
Comment (jameswhollister) [reply src/renderer/services/dex.js]: I don't believe any of the error messages in js files are localized, this should be a separate ticket. I don't see an example of how to import the localization into a service.

**Pull request merged**